source/extensions/filters/http/ext_authz/ext_authz.cc

```diff
@@ -94,35 +95,59 @@ void Filter::onComplete(Filters::Common::ExtAuthz::CheckStatus status) {
     break;
   case CheckStatus::Denied:
     cluster_->statsScope().counter("ext_authz.denied").inc();
-    Http::CodeUtility::ResponseStatInfo info{config_->scope(),
-                                             cluster_->statsScope(),
-                                             EMPTY_STRING,
-                                             enumToInt(Http::Code::Forbidden),
-                                             true,
-                                             EMPTY_STRING,
-                                             EMPTY_STRING,
-                                             EMPTY_STRING,
-                                             EMPTY_STRING,
-                                             false};
+    Http::CodeUtility::ResponseStatInfo info{
+        config_->scope(),
+        cluster_->statsScope(),
+        EMPTY_STRING,
+        (update_status_code ? response->status_code : enumToInt(Http::Code::Forbidden)),
+        true,
+        EMPTY_STRING,
+        EMPTY_STRING,
+        EMPTY_STRING,
+        EMPTY_STRING,
+        false};
     Http::CodeUtility::chargeResponseStat(info);
     break;
   }
 
   // We fail open/fail close based of filter config
   // if there is an error contacting the service.
-  if (status == CheckStatus::Denied ||
-      (status == CheckStatus::Error && !config_->failureModeAllow())) {
-    Http::HeaderMapPtr response_headers{new Http::HeaderMapImpl(*getDeniedHeader())};
-    callbacks_->encodeHeaders(std::move(response_headers), true);
+  if (response->status == CheckStatus::Denied ||
+      (response->status == CheckStatus::Error && !config_->failureModeAllow())) {
+    Http::HeaderMapPtr response_headers;
+    if (update_status_code) {
+      response_headers = std::make_unique<Http::HeaderMapImpl>(Http::HeaderMapImpl{
+          {Http::Headers::get().Status, std::to_string(response->status_code)}});
+    } else {
+      response_headers = std::make_unique<Http::HeaderMapImpl>(*getDeniedHeader());
+    }
+
+    for (const auto& header : response->headers) {
+      response_headers->setReferenceKey(header.first, header.second);
+    }
+
+    if (response->body) {
+      callbacks_->encodeHeaders(std::move(response_headers), false);
+      callbacks_->encodeData(*response->body.get(), true);
+    } else {
+      callbacks_->encodeHeaders(std::move(response_headers), true);
+    }
```

Comment:
may be wrapping the encoding of response headers and body into a function would help with parsing of the logic here.

Reply:
I think so too. I will try to simplify this logic, but first I will make sure that we have tests for all the permutations.